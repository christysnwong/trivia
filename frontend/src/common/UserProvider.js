import React, {useEffect, useState} from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "test@test.com",
  stats: {
    level: 14,
    title: "Pro",
    "quizzes completed": 34,
    points: 3440,
    remainingPts: 60,
    levelPts: 300,
  },
};

const UserProvider = ({
  children,
  currUser = demoUser
}) => {
  
  const [currentUser, setCurrUser] = useState();

  return (
    // <UserContext.Provider value={{ currUser }}>
    <UserContext.Provider value={{ currUser, setCurrUser }}>
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
