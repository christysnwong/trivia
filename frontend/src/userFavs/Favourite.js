import { useState, useEffect, useContext } from "react";
import TriviaApi from "../api/api";

import UserContext from "../common/UserContext";
import FavTable from "./FavTable";

import CreateFolderForm from "./CreateFolderForm";
import EditFolderForm from "./EditFolderForm";

import { Collapse } from "reactstrap";

import Noty from "noty";

const Favourite = () => {
  const { currUser } = useContext(UserContext);

  const [userFolders, setUserFolders] = useState([]);
  const [folder, setFolder] = useState({});
  const [fav, setFav] = useState([]);

  const [createFormIsOpen, setCreateFormIsOpen] = useState(false);
  const createFormToggle = () => setCreateFormIsOpen(!createFormIsOpen);

  const [editFormIsOpen, setEditFormIsOpen] = useState(false);
  const [folderIdToRename, setFolderIdToRename] = useState(0);
  const editFormToggle = () => setEditFormIsOpen(!editFormIsOpen);

  const [folderIdToDelete, setFolderIdToDelete] = useState(0);

  // load selected folder's trivia and display on page
  const getFolder = (evt) => {

    TriviaApi.getFolderTrivia(
      currUser.username,
      evt.target.parentElement.id
    ).then((f) => {

      setFolder((folder) => ({
        ...folder,
        id: f.folderId,
        name: f.folderName,
      }));

      setFav(f.trivia);
    });
  };

  // toggle edit form to show, get folder id for folder to be renamed
  const renameFolder = (evt) => {
    editFormToggle();

    setFolderIdToRename(+evt.target.closest("div").id);
  };

  // delete a folder
  const removeFolder = (evt) => {
    
    try {
      let folderId = folderIdToDelete;

      TriviaApi.removeFolder(currUser.username, folderId).then((res) => {
        if (res.deleted) {
          new Noty({
            type: "success",
            layout: "topRight",
            theme: "bootstrap-v4",
            text: "This folder is successfully deleted.",
            timeout: "3000",
            progressBar: true,
            closeWith: ["click"],
            killer: true,
          }).show();

          setUserFolders((userFolders) =>
            userFolders.filter((f) => f.folderId !== folderId)
          );

          if (folderId === folder.id) {
            setFolder({});
            setFav([]);
          }
        }
      });
    } catch (e) {
      console.error(e);

      new Noty({
        type: "error",
        layout: "topRight",
        theme: "bootstrap-v4",
        text: e,
        timeout: "3000",
        progressBar: true,
        closeWith: ["click"],
        killer: true,
      }).show();

    }
  };

  // delete a trivia from a folder
  const removeTrivia = (evt) => {

    try {
      let triviaId = +evt.target.closest("tr").id;

      TriviaApi.removeTrivia(currUser.username, triviaId).then((res) => {

        if (res.deleted) {

          new Noty({
            type: "success",
            layout: "topRight",
            theme: "bootstrap-v4",
            text: "This trivia is successfully deleted.",
            timeout: "3000",
            progressBar: true,
            closeWith: ["click"],
            killer: true,
          }).show();

          setFav((fav) => fav.filter((entry) => entry.id !== triviaId));
        }
      });
    } catch (e) {
      console.error(e);

      new Noty({
        type: "error",
        layout: "topRight",
        theme: "bootstrap-v4",
        text: e,
        timeout: "3000",
        progressBar: true,
        closeWith: ["click"],
        killer: true,
      }).show();

    }
    
  };

  // move trivia to another folder
  const moveTrivia = (evt) => {

    try {
      let triviaId = +evt.target.closest("tr").id;
      let folderName = evt.target.innerText;

      TriviaApi.moveTrivia(currUser.username, triviaId, {
        userId: currUser.id,
        folderName,
      }).then((res) => {

        if (res.updated) {

          new Noty({
            type: "success",
            layout: "topRight",
            theme: "bootstrap-v4",
            text: `This trivia is successfully moved to folder ${folderName}`,
            timeout: "3000",
            progressBar: true,
            closeWith: ["click"],
            killer: true,
          }).show();

          setFav((fav) => fav.filter((entry) => entry.id !== triviaId));
        }
      });
    } catch (e) {
      console.error(e);

      new Noty({
        type: "error",
        layout: "topRight",
        theme: "bootstrap-v4",
        text: e,
        timeout: "3000",
        progressBar: true,
        closeWith: ["click"],
        killer: true,
      }).show();
    }
    
  };

  // get all user's folders
  useEffect(() => {
    TriviaApi.getAllFolders(currUser.username).then((folders) => {
      setUserFolders(folders);
    });
  }, []);

  return (
    <div className="col-10 offset-1 col-lg-8 offset-lg-2 col-xxl-6 offset-xxl-3 col-xxxl-4 offset-xxxl-4">

      <h3 className="mb-5">
        <i className="fa-solid fa-heart me-2"></i>
        My Favourites
      </h3>

      <h3 className="mb-4">
        Folders
        <span className="ms-2 pointer-cursor" onClick={createFormToggle}>
          <i className="fa-solid fa-folder-plus"></i>
        </span>
      </h3>

      {/* create form for creating a folder */}
      <Collapse isOpen={createFormIsOpen}>
        <CreateFolderForm
          setUserFolders={setUserFolders}
          setCreateFormIsOpen={setCreateFormIsOpen}
        />
      </Collapse>

      <div className="row row-cols-3 mb-4">
        {userFolders.map((f) => (
          <div className="col mb-3 d-flex align-items-stretch" key={f.folderId}>
            <div className="card w-100">
              <div className="card-body" id={f.folderId}>
                <span className="card-text pointer-cursor" onClick={getFolder}>
                  {f.name}
                </span>
                {f.name !== "All" && (
                  <>
                    <span
                      className="ms-2 float-end pointer-cursor"
                      data-bs-toggle="modal"
                      data-bs-target="#deleteModal"
                      onClick={(evt) =>
                        setFolderIdToDelete(+evt.target.closest("div").id)
                      }
                    >
                      <i className="fa-sharp fa-solid fa-trash"></i>
                    </span>
                    <span
                      className="float-end pointer-cursor"
                      onClick={renameFolder}
                    >
                      <i className="fas fa-pen-alt"></i>
                    </span>
                  </>
                )}
              </div>
            </div>
          </div>
        ))}
      </div>

      {/* modal for deleting a folder */}
      <div
        className="modal fade"
        id="deleteModal"
        tabIndex="-1"
        aria-labelledby="deleteModalLabel"
        aria-hidden="true"
      >
        <div className="modal-dialog">
          <div className="modal-content">
            <div className="modal-header">
              <h5 className="modal-title" id="deleteModalLabel">
                Deleting Folder
              </h5>
              <button
                type="button"
                className="btn-close"
                data-bs-dismiss="modal"
                aria-label="Close"
              ></button>
            </div>
            <div className="modal-body">
              Are you sure that you want to delete this folder? Any trivia
              stored in this folder will also be deleted as well.
            </div>
            <div className="modal-footer">
              <button
                type="button"
                className="btn btn-secondary"
                data-bs-dismiss="modal"
              >
                Close
              </button>
              <button
                type="button"
                className="btn btn-danger"
                onClick={removeFolder}
                data-bs-dismiss="modal"
              >
                Delete
              </button>
            </div>
          </div>
        </div>
      </div>

      {/* edit form for renaming a folder */}
      <Collapse isOpen={editFormIsOpen}>
        {userFolders.length > 0 && folderIdToRename && (
          <EditFolderForm
            folderId={folderIdToRename}
            folderName={
              userFolders.filter((f) => f.folderId === folderIdToRename)[0]
                ?.name ?? ""
            }
            setUserFolders={setUserFolders}
            setEditFormIsOpen={setEditFormIsOpen}
          />
        )}
      </Collapse>

      <div className="row">
        <div className="col-12">
          <FavTable
            userFolders={userFolders.filter((f) => f.folderId !== folder.id)}
            folder={folder}
            entries={fav}
            removeTrivia={removeTrivia}
            moveTrivia={moveTrivia}
          />
        </div>
      </div>
    </div>
  );
};

export default Favourite;
