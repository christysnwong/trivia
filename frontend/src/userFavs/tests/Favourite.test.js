import React from "react";
import { render, screen, fireEvent, act } from "@testing-library/react";
import Favourite from "../Favourite";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "../../common/UserProvider";

import TriviaApi from "../../api/api";

import { commonBeforeEach, commonAfterAll } from "./_testCommonUserFavs";

beforeEach(commonBeforeEach);
afterAll(commonAfterAll);

it("renders without crashing", function () {
  render(
    <UserProvider>
      <Favourite />
    </UserProvider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <Favourite />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("checks if TriviaApi.getAllFolders is called", async () => {
  jest.spyOn(React, "useEffect").mockImplementation((f) => f());
  jest.spyOn(TriviaApi, "getAllFolders");
  render(
    <UserProvider>
      <Favourite />
    </UserProvider>
  );

  expect(TriviaApi.getAllFolders).toHaveBeenCalled();

});

it("loads folders and trivia", async function() {
  jest.spyOn(React, "useEffect").mockImplementation((f) => f());
  jest.spyOn(TriviaApi, "getAllFolders");
  jest.spyOn(TriviaApi, "getFolderTrivia");
  render(
    <UserProvider>
      <Favourite />
    </UserProvider>
  );

  expect(await screen.findByText(/all/i)).toBeInTheDocument();
  expect(await screen.findByText(/test1/i)).toBeInTheDocument();
  expect(await screen.findByText(/test2/i)).toBeInTheDocument();
  expect(await screen.findByText(/test3/i)).toBeInTheDocument();

  const allFolder = await screen.findByText(/all/i);

  fireEvent.click(allFolder);

  expect(TriviaApi.getFolderTrivia).toHaveBeenCalled();

})
