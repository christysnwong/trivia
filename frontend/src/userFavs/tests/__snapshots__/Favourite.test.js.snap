// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-10 offset-1 col-lg-8 offset-lg-2 col-xxl-6 offset-xxl-3 col-xxxl-4 offset-xxxl-4"
  >
    <h3
      class="mb-5"
    >
      <i
        class="fa-solid fa-heart me-2"
      />
      My Favourites
    </h3>
    <h3
      class="mb-4"
    >
      Folders
      <span
        class="ms-2 pointer-cursor"
      >
        <i
          class="fa-solid fa-folder-plus"
        />
      </span>
    </h3>
    <div
      class="collapse"
    >
      <div
        class="card mb-5"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="mb-2"
            >
              <label
                class="form-label"
                for="username"
              >
                Create a Folder: 
              </label>
              <input
                class="form-control"
                id="folderName"
                maxlength="20"
                name="folderName"
                placeholder="New Folder Name"
                value=""
              />
            </div>
            <button
              class="btn btn-secondary mt-4 me-3"
              method="get"
              type="button"
            >
              Cancel
            </button>
            <button
              class="btn btn-primary mt-4"
              method="post"
              type="submit"
            >
              Create
            </button>
          </form>
        </div>
      </div>
    </div>
    <div
      class="row row-cols-3 mb-4"
    />
    <div
      aria-hidden="true"
      aria-labelledby="deleteModalLabel"
      class="modal fade"
      id="deleteModal"
      tabindex="-1"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
              id="deleteModalLabel"
            >
              Deleting Folder
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              data-bs-dismiss="modal"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            Are you sure that you want to delete this folder? Any trivia stored in this folder will also be deleted as well.
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-secondary"
              data-bs-dismiss="modal"
              type="button"
            >
              Close
            </button>
            <button
              class="btn btn-danger"
              data-bs-dismiss="modal"
              type="button"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="collapse"
    />
    <div
      class="row"
    >
      <div
        class="col-12"
      >
        <div>
          <p>
            You can either create a new folder by clicking on 
            <i
              class="fa-solid fa-folder-plus"
            />
             or select a folder above to view your favourited trivia.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
